File: Data Structures & Algorithms/foreign-dictionary/submission-0.py
class Solution:
    def foreignDictionary(self, words: List[str]) -> str:
        adj = {}
        #ADJ LIST
        for w in words:
            for c in w:
                adj[c] = set()

        for i in range(len(words) - 1):
            word1 = words[i]
            word2 = words[i + 1]
            minLen = min(len(word1), len(word2))
            if len(word1) > len(word2) and word1[:minLen] == word2[:minLen]:
                return ""
            for j in range(minLen):
                if word1[j] != word2[j]:
                    adj[word1[j]].add(word2[j])
                    break
            
        visited = {}

        res = []

        def dfs(char):
            if char in visited:
                return visited[char]

            visited[char] = True

            for neighbor in adj[char]:
                if dfs(neighbor):
                    return True

            visited[char] = False
            res.append(char)
        
        for char in adj:
            if dfs(char):
                return ""
        res.reverse()
        return "".join(res)

File: Data Structures & Algorithms/edit-distance/submission-0.cpp
class Solution {
public:
    int minDistance(string word1, string word2) {
        int m = word1.length(), n = word2.length();
        vector<vector<int>> dp(m + 1, vector<int>(n + 1, 0));
        
        // Base cases
        for (int i = 0; i <= m; i++) {
            dp[i][0] = i;
        }
        for (int j = 0; j <= n; j++) {
            dp[0][j] = j;
        }


        for (int i = 0; i <= m; i++) {
            for (int j = 0; j <= n; j++) {
                if (i == 0) {
                    dp[i][j] = j;
                } 
                else if (j == 0) {
                    dp[i][j] = i; 
                }
                else if(word1[i-1] == word2[j-1]){
                    dp[i][j] = dp[i-1][j-1];
                }else{
                    dp[i][j] = min({dp[i-1][j],    // Delete
                                    dp[i][j-1],    // Insert  
                                    dp[i-1][j-1]}) // Replace
                                    + 1;
                }
            }
        }
        return dp[m][n];

    }
};
